course/viewbase.py

```diff
@@ -116,6 +123,15 @@ def get_course_instance_object(self):
             course__url=self.kwargs[self.course_kw],
         )
 
+    def handle_exception(self, exc):
+        if isinstance(exc, TranslationNotFound):
+            instance_languages = self.instance.language.strip("|").split("|")
+            url = remove_query_param_from_url(self.request.get_full_path(), 'hl')
+            for i, lang in enumerate(instance_languages):
+                instance_languages[i] = {"name": get_language_info(lang)['name'], "url": update_url_params(url, {'hl' : lang})}
```

Comment:
Hmm.. If A+ would be configured to use languages `en-fi` and `fi-fi` and the course would have `en` and `fi`, then this would generate links, which wouldn't use the correct value for the `hl` parameter. So, you probably need to iterate over `settings.LANGUAGES` and filter that, so that only those which have translation in the course are listed.

Reply:
Can you clarify why is the `settings.LANGUAGES` needed? If the course has languages `en` and `fi`, then the links would have `?hl=en` and `?hl=fi`, aren't those the right ones?

Reply:
Or should it be `?hl=en-fi` and `?hl=fi-fi`?

Reply:
Your other part of the code requires that the `?hl=<lang>` is found from the `settings.LANGUAGES`, thus `?hl=en` should result 404 if there is only `en-us` and `fi`.

Reply:
Maybe it's still quite unclear for me how the `en-fi` should work. If A+ has language `en-fi` and course is available in `en`, the course content should still surely be displayed in English? And I guess `hl` tag can only contain the first two characters of  the language code or how should it work with `en-us` for example?

Reply:
As the tanslation.activate is called with the value of the `hl`, then the `hl` should contain the full language info (i.e. locale = language + location). The second code, in course instance, then shortens this locale to just tlanguage, which are used to select content from the course.

In other words, all Django stuff expect the language parameter to be fully qualified name (e.g. `en-us`), so it can select correct translation file and formats. However, the course content is translated only with 2 letter codes, so we need to shorten the `en-us` to just `en` in all course material related places (and assessment tools, like grading).

Reply:
Thanks, that clears some things up but I'm somewhat confused about the formats still.

This is more a concern of https://github.com/apluslms/a-plus/pull/604, but that depends on how we solve the language locale/region issues here:

If we added the `formats/en-fi/formats.py` folder and it would be possible for the language to be `en`, `en-fi` and even `en-us` or `en-gb` all in the same session, would the date and time formats then vary between the ones defined in `formats/en/formats.py` and `formats/en-fi/formats.py`? That would be very confusing.

I don't see the point in supporting these different language formats (other than maybe for displaying dates and times) because it's not even possible for users to pick `en-fi`, `en-gb` or `en-us` as the preferred language, there is only `English`, so in which case `en-fi` would come up?

Reply:
Maybe it would be simpler if the language could be only `fi` or `en` and there was a separate `date_format` field in `UserProfile` that could be changed from profile page, but I'm not sure how displaying the correct format could be implemented.

Reply:
Let's write this down. Back in the day at times of UNIX, POSIX standard specified [locale syntax](https://en.wikipedia.org/wiki/Locale_(computer_software)) `[language[_territory][.codeset][@modifier]]`, e.g., `en_US.UTF-8` and `fi_FI.UTF-8` (the charset is now-a-days pretty useless as unicode contains it all, things were different before it). Another relevant info is how [HTTP specified](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language) `Accept-Language` header (e.g. `en`, `en-US`, `fi`) (see also [BCP 47 / IETF language tag](https://en.wikipedia.org/wiki/IETF_language_tag)). The problem which people has tried to solve is how to differentiate differences in the language.

A good question is why formats and languages are typically connected like this. Like why `fi_FI` defines formats used in Finalnd? Technically, that string contains all required information. The `en_FI` could be parsed as show texts with language English, but use geographical information related to Finland. For the latter, there is clear definition how numbers and such are represented in our location. Interestingly in Switzerland, the location part is not enough to define the used formats, thus both the language and the location are required to get the exact information.

Many distributions don't include all possible combinations for locales, e.g., `en_FI` doesn't exist in Debian or Ubuntu. However, it's part of the [Unicode Common Locale Data Repository](http://cldr.unicode.org/). This was the best reference to say that it's a valid locale string, which I could find.

So, languages _often_ requires the geographical location to represent the whole UI correctly (including formats).

However, we have so far went with the simpler language formats for course material. Though, I think we should start to plan that the course language code could include possibly include the location information too, at least to some extend. The full IETF standard would be very complicated though. So, for now we use only the two letter code to indicate the language for the material. As the locale to language map is always surjection (i.e. for every language, there can exist one or more locales). Thus, we can always map any locale to a language, but we can't map any language to one locale.

That said, Django should need to provide those formats files, as that information should be available via some library, but apparently that isn't the case yet. Thus, for the formats, I think we need to provide definitions for those locales, which we would like to use, but are not covered by the Django yet.

Next problem. The `hl` parameter (and userprofile.language and enrollment.language) requires the fully qualified locale name always, so the `translation.activate()` can operate correctly, as it needs to find the correct formats.py file too. If we think A+ as an application, which is used to read course material, then we can think that the application is always configured to show the UI with specific locale and not just with specific language. However, the application would select the best matching material from the database for that locale.

I guess, the biggest challenges are the cases when  the best information comes from `Accept-Language` and it doesn't include the location (e.g. just `en`). Maybe that is when we need to guess the best possible language, and provide way to change it.

Another annoyance would be that when viewing the material in English (US), we could provide button to change the language to English (UK), but as that would so the same content, but only change how the datetimes are represented, it could be very confusing to the user. I'm not sure what would be the best. Luckily, it's likely that there exist only one English definition in the settings, which would mean that typically locales and languages would have one to one mapping.

Maybe this contains some valuable information, maybe not. Have to say that things didn't become super clear for me either, even though I spend the evening reading about this stuff.

To summarise, the application (A+) should use the most accurate locale information (e.g. `en-us`, `en-fi`, `fi-fi`), when possible, and would select the best matching course material. If no version of the material is available for the locale, then A+ would show 404 with the list of possible versions (even if there would be multiple English variants).

Reply:
Thank you for the clarification, I still have a few questions about the logic/implementation:
1. Do we want to change the `LANGUAGES` in settings.py to contain only `('en-fi', 'English (FI)'), ('fi-fi', 'Finnish (FI)')` (and maybe also 'en-gb' and 'en-us') or just add those two and keep `('en', 'English'), ('fi', 'Finnish')` there too? If we replace the shorter formats completely in this PR, the `formats.py` files probably need to be replaced here also.
2. A course is only available in English, and user tries to access it in Finnish, what languages are there in the 404 error page, i.e. is there only `English` as the course language does not contain information about location or is there `English (FI)`, `English (UK)` etc?
3. When using only `?hl=en`, do we want to show the error page or direct the user to `?hl=en-fi`? I think it would be better to not show the error but the logic might get complicated if we need to guess the longer language code from the shorter one and change it every time. 
4. How about with Finnish? When we probably have only `fi-fi`, it would make sense to accept both `fi` and `fi-fi` as language codes and not show the error page when query parameter is `?hl=fi`.

There's so many unclear things about this that we need to find some compromise solution in scope of this PR if we want to get this somehow done and merged.

Reply:
I think that `('en', 'English'), ('fi', 'Finnish')` has to be there as long as course code can be only 2 characters. Then we might be left with this kind of error message:
![404-english](https://user-images.githubusercontent.com/36920208/91039334-10df6900-e615-11ea-9461-7c5f825bd8fc.png)

Reply:
> Do we want to change the LANGUAGES in settings.py to contain only ('en-fi', 'English (FI)'), ('fi-fi', 'Finnish (FI)') (and maybe also 'en-gb' and 'en-us') or just add those two and keep ('en', 'English'), ('fi', 'Finnish') there too? If we replace the shorter formats completely in this PR, the formats.py files probably need to be replaced here also.

To help with #604, I would say we want to have `en-fi` and `fi` (`fi-fi` doesn't add any more information).

> A course is only available in English, and user tries to access it in Finnish, what languages are there in the 404 error page, i.e. is there only English as the course language does not contain information about location or is there English (FI), English (UK) etc?

They should see list of all possible languages, which would match, i.e., `English (US), English (FI), Finnish` for a course with langs `['en', 'fi']` and LANGUAGES `['en-fi', 'en-us', 'fi']`, when they try to access the course with some other language. This isn't perfect, but I think the best we could do.

Better option would be having a different UI language and course language, but that is not the case for current design. However, I think in most of the cases, the system will be configured with a single English locale.

> When using only ?hl=en, do we want to show the error page or direct the user to ?hl=en-fi? I think it would be better to not show the error but the logic might get complicated if we need to guess the longer language code from the shorter one and change it every time.

Technically, we should show an error, but as you said that it would be annoying. Hence, let's make an educated guess by redirecting the user to the first matching locale in the LANGUAGES list (it's an ordered list).

> How about with Finnish? When we probably have only fi-fi, it would make sense to accept both fi and fi-fi as language codes and not show the error page when query parameter is ?hl=fi.

I think it should be sufficient to use just `fi` in the defined languages (it's fully qualified name, as there is no sub languages. Related languages have their own tags, e.g., `fit` for [Tornedalen Finnish](https://en.wikipedia.org/wiki/Me%C3%A4nkieli)).

Thus, we probably should also do lookup for a shorter `hl` entry too. I.e., if we have `?hl=tag1-tag2-tag3-Tag4`, we would first check if there is `tag1-tag2-tag3-Tag4` in LANGUAGES, if not then check `tag1-tag2-tag3`, `tag1-tag2` and finally `tag1`. And as always, select the first from the list of matches and redirect.

As a last remark, I would expect the settings to contain something like this:

```python
LANGUAGES = (
  ('en-FI', _('English')),
  ('fi', _('Finnish')),
  ('sv', _('Swedish')),
)
```

Or in the US:

```python
LANGUAGES = (
  ('en-US', _('English')),
)
```

Or in the Mexico (I guess):

```python
LANGUAGES = (
  ('es-MX', _('Spanish')),
  ('en-US', _('English')),
)
```

p.s. the region code should be styled with uppercase letters, but I'm not sure how Django handles that in this settings value. The `translation.activate` handles it correctly, if I remember right.

These details (i.e. `es-MX` vs. `es`) only help with formats in our case for now.

Reply:
Your suggestion seems otherwise good, but I'm a bit worried whether things are going to break if we remove `en` from `settings.py` when it's still used as course instance language. I tested that today with the test database and it deleted the `en` from language field in course instances where it was used, I suppose because it wasn't supported in `settings.py`.
We could have `English (US)` or `English (UK)` with language code `en` as the default, and option to change it to `English (FI)` at least from profile page, which then allows to use the Finnish formats.
What do you think of this? 
```
LANGUAGES = [
    ('en-FI', 'English (FI)'),
    ('en', 'English (US)'),
    ('fi', 'Finnish'),
]
```

Reply:
I think we have to have a discussion in a meeting about these language code issues (on Tuesday?). I'm afraid @raphendyr is making this really complicated and I hope it wouldn't need to be so. I can see this is not "as easy as it looks" and that languages and localizations have many twists. However, I would like to consider whether we could still aim for a simple, minimal approach. That is, language choice would focus on the language, not the region, and short 2-character language codes would be used as much as possible. How about continuing with this after the daily?

Reply:
@Mankro Sounds good for me.

Reply:
Yes, rest of the software doesn't yet handle language codes longer than 2 characters. I would like to get this one part working, so rest of the components can be fixed while moving forward. In the end, there are only few places which need to be fixed.

> However, I would like to consider whether we could still aim for a simple, minimal approach. That is, language choice would focus on the language, not the region, and short 2-character language codes would be used as much as possible.

I wish it was so simple. We are just mixing locale information and course languages to the same variable, where the course language doesn't often include enough information for the rest of the system.

Even ISO 639-3 codes (3 letter language codes) would already work for many cases, but not all, e.g., `pt-BR` (though, I have no idea why it doesn't have ISO code).

Reply:
We made a decision to limit the language strings to 2 character codes (i.e. ISO 639-2 codes), which should work for the currently known use cases and has less change to break stuff. The code should in most cases ensure the longer or invalid codes are cut or dropped, so that only valid codes are used.